File: cpp/1019. 链表中的下一个更大节点.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        stack<pair<int,int> > st;
        vector<int> ans;
        ListNode* p=head;
        int ind=-1;
        while(p!=nullptr){
            ++ind;
            ans.push_back(0);
            while(!st.empty() && st.top().first<p->val){
                ans[st.top().second]=p->val;
                st.pop();
            }
            st.push({p->val,ind});
            p=p->next;
        }
        return ans;
    }
};

File: 刷题/golang/1019. 链表中的下一个更大节点.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func nextLargerNodes(head *ListNode) []int {
	var ans []int
	var stack [][]int
	cur := head
	ind := -1
	for cur != nil {
		ind++
		ans = append(ans, 0)
		for len(stack) > 0 && stack[len(stack)-1][0] < cur.Val {
			ans[stack[len(stack)-1][1]] = cur.Val
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, []int{cur.Val, ind})
		cur = cur.Next
	}
	return ans
}